docs(product): document DuckDB as a data source (#6969)
cube-js/cube: docs/docs-new/pages/product/configuration/data-sources/duckdb.mdx

---
redirect_from:
  - /config/databases/aws-athena

> ---

Reviewer: Why?

# DuckDB

> [duckdb] is an in-process SQL OLAP database management system, and has support

Reviewer: ```suggestion
[DuckDB][duckdb] is an in-process SQL OLAP database management system, and has support
```

> [duckdb] is an in-process SQL OLAP database management system, and has support

Reviewer: Let's also mention that DuckDB can serve as a lightweight alternative to other query engines like AWS Athena or Trino.

> for querying data in CSV, JSON and Parquet formats from an AWS S3 bucket.

Reviewer: DuckDB actually works with [S3, GCS, and Cloudflare R2](https://duckdb.org/docs/guides/import/s3_import#how-to-load-a-parquet-file-directly-from-s3-gcs-or-r2). Let's make this docs page less S3-centric.

## Prerequisites

- [A set of IAM credentials][aws-docs-access-keys] which allow access to [AWS
  S3][aws-s3]
- [The AWS region][aws-docs-regions]
- [The S3 bucket][aws-s3] on AWS to read data from

## Setup

### Manual

Add the following to a `.env` file in your Cube project:

```dotenv
CUBEJS_DB_TYPE=duckdb
```

### Cube Cloud

In Cube Cloud, select **DuckDB** when creating a new deployment and fill in the
required fields:

<InfoBox>

If you are not using MotherDuck, leave the **MotherDuck Token** field blank.

</InfoBox>

<Screenshot
  alt="Cube Cloud DuckDB Configuration Screen"
  src="https://ucarecdn.com/ffad02b1-a563-47e4-bbc0-0302ab046ccd/"
/>

> ## Environment Variables

Reviewer: Please review and fix all env vars in this section.

| `CUBEJS_DUCKDB_S3_ACCESS_KEY_ID`        | The AWS Access Key ID to use for database connections                               | A valid AWS Access Key ID                                     |    ✅    |                                    ✅                                    |
| `CUBEJS_DUCKDB_S3_SECRET_ACCESS_KEY_ID` | The AWS Secret Access Key to use for database connections                           | A valid AWS Secret Access Key                                 |    ✅    |                                    ✅                                    |
| `CUBEJS_DUCKDB_S3_ENDPOINT`             | The AWS S3 endpoint                                                                 | [A valid AWS S3 endpoint][aws-docs-vpc-endpoints]             |    ✅    |                                    ✅                                    |
| `CUBEJS_DUCKDB_S3_REGION`               | The AWS region of the Cube deployment                                               | [A valid AWS region][aws-docs-regions]                        |    ✅    |                                    ✅                                    |

> | `CUBEJS_CONCURRENCY`                    | The number of concurrent connections each queue has to the database. Default is `2` | A valid number                                                |    ❌    |                                    ❌                                    |

Reviewer 2: https://github.com/cube-js/cube/pull/6982

`CUBEJS_DB_DUCKDB_MEMORY_LIMIT` - 1gb/10gb/1mb - by default it's 75% of RAM

# DuckDB

[DuckDB][duckdb] is an in-process SQL OLAP database management system, and has

> support for querying data in CSV, JSON and Parquet formats from an S3-compatible
> API. This means you can [query data stored in AWS S3 or CloudFlare

Reviewer: ```suggestion
support for querying data in CSV, JSON and Parquet formats from an AWS S3-compatible
blob storage. This means you can [query data stored in AWS S3, Google Cloud Storage, or CloudFlare
```

R2][duckdb-docs-s3-import].

> Cube can also connect to [MotherDuck][motherduck], a cloud-based serverless
> analytics platform built on DuckDB.

Reviewer: ```suggestion
Cube can also connect to [MotherDuck][motherduck], a cloud-based serverless
analytics platform built on DuckDB. When connected to MotherDuck, DuckDB uses [hybrid execution](https://motherduck.com/docs/architecture-and-capabilities#hybrid-execution)
and routes queries to S3 through MotherDuck for better performance. 
```

| `CUBEJS_DB_DUCKDB_MOTHER_DUCK_TOKEN`       | The service token to use for connections to MotherDuck                                                                | A valid [MotherDuck service token][motherduck-docs-svc-token] |    ❌    |                                    ✅                                    |
| `CUBEJS_DB_DUCKDB_S3_ACCESS_KEY_ID`        | The Access Key ID to use for database connections                                                                     | A valid Access Key ID                                         |    ❌    |                                    ✅                                    |
| `CUBEJS_DB_DUCKDB_S3_SECRET_ACCESS_KEY_ID` | The Secret Access Key to use for database connections                                                                 | A valid Secret Access Key                                     |    ❌    |                                    ✅                                    |

> | `CUBEJS_DB_DUCKDB_S3_ENDPOINT`             | The S3 endpoint                                                                                                       | A valid S3 endpoint                                           |    ✅    |                                    ✅                                    |

Reviewer: ```suggestion
| `CUBEJS_DB_DUCKDB_S3_ENDPOINT`             | The S3 endpoint                                                                                                       | A valid [S3 endpoint](https://duckdb.org/docs/guides/import/s3_import) |    ✅    |                                    ✅                                    |
```

> | `CUBEJS_DB_DUCKDB_S3_REGION`               | The region of the bucket                                                                                              | A valid AWS region                                            |    ✅    |                                    ✅                                    |

Reviewer: ```suggestion
| `CUBEJS_DB_DUCKDB_S3_REGION`               | The [region of the bucket](https://duckdb.org/docs/guides/import/s3_import) | A valid AWS region                                            |    ✅    |                                    ✅                                    |
```
